src/main.rs

```diff
@@ -175,6 +193,53 @@ fn list_cmd(hidapi: &HidApi) {
     }
 }
 
+pub fn tap_command_loop(device: &HidDevice) -> Result<()> {
+    const HISTORY_NAME: &str = ".bose-dfu_history";
+
+    let mut rl = DefaultEditor::new()?;
+    let history_file = match env::home_dir() {
+        Some(home_path) => home_path.join(HISTORY_NAME),
+        None => PathBuf::from(HISTORY_NAME),
+    };
+    if rl.load_history(&history_file).is_err() {
+        println!("No previous history.");
+    }
+
+    loop {
+        let readline = rl.readline("> ");
+        match readline {
+            Ok(line) => {
+                if line.is_empty() {
+                    continue;
+                } else if line == "." {
+                    break;
+                }
+
```

Comment:
```suggestion

                rl.add_history_entry(line.as_str())?;
```

I think this is why you're not getting history.